# Imports

In [53]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import shutil as st

from opendata import OpenData
from opendata.models import LocalAthlete
from lmfit import minimize, Parameters
from dateutil.parser import parse
import datetime
from tqdm import tqdm

# Create OpenData object

In [54]:
od = OpenData()

# Taking all athletes

In [3]:
athletes = list(od.remote_athletes())

In [ ]:
len(athletes)

# Read athletes.csv file to fast find who has no lest than 200 activities and then collect their data localy

In [ ]:
AthleteDF = pd.read_csv('athletes.csv')
AthleteDF

# Store locally athletes with no less then 200 bike activities

Too long way...

In [ ]:
def filter_athletes(athletes):
    filtered_athletes = []
    for athlete in tqdm(athletes):
        i = 0
        try:
            activities = list(athlete.activities())
        except:
            continue
        try:
            for activ in activities:
                if (activ.metadata['sport'] == 'Bike'):
                    i = i + 1
                if (i >= 100):
                    filtered_athletes.append(athlete)
                    athlete.store_locally()
                    break
        except:
            continue
    return filtered_athletes

In [ ]:
filtered_athletes = filter_athletes(athletes)
len(filtered_athletes)

A bit clever trick:

In [ ]:
def download_athletes(DataFrame):
    athletes = []
    for i, row in tqdm(DataFrame.iterrows()):
        if (row['bike'] >= 200 and row['run'] < 10 and row['swim'] < 10):
            athlete = row['id']
            athletes.append(athlete)
    return athletes

In [ ]:
interesting_athletes = download_athletes(AthleteDF)
len(interesting_athletes)

In [ ]:
for athlete_id in tqdm(interesting_athletes):
    if athlete_id not in local_athletes:
        athlete = od.get_remote_athlete(athlete_id)
        athlete.store_locally()

In [ ]:
local_athletes = list(od.local_athletes())
len(local_athletes)

# Choosing random athlete

In [4]:
athlete = athletes[448]

In [5]:
athlete.metadata['ATHLETE']

{'gender': 'M', 'yob': '1979', 'id': '{11bf3ef3-c7b3-4d8d-be36-9021982b51f3}'}

In [6]:
athlete.id

'11bf3ef3-c7b3-4d8d-be36-9021982b51f3'

In [7]:
activities = list(athlete.activities())

In [ ]:
activities

# Choosing random activity

In [8]:
n = 154

In [9]:
activities[n].metadata

{'date': '2018/11/05 15:26:21 UTC',
 'data': 'TDSPHC-AGL-----',
 'sport': 'Bike',
 'METRICS': {'a_skiba_xpower': ['198.90497', '3131.00000'],
  'a_skiba_relative_intensity': ['0.84640', '3131.00000'],
  'a_skiba_bike_score': '62.30690',
  'a_skiba_variability_index': '1.18095',
  'a_skiba_response_index': '1.47541',
  'a_coggan_np': ['206.86499', '3131.00000'],
  'a_coggan_if': ['0.88028', '3131.00000'],
  'a_coggan_tss': '67.39363',
  'a_coggam_variability_index': ['1.22821', '3131.00000'],
  'a_friel_efficiency_factor': '1.53446',
  'a_coggan_tssperhour': ['77.48868', '0.86972'],
  'aerobic_decoupling': '11.65664',
  'power_index': ['63.33773', '1.00000'],
  'ride_count': '1.00000',
  'workout_time': '3131.00000',
  'time_riding': '3124.00000',
  'total_distance': '28.80733',
  'climb_rating': ['0.17302', '1.00000'],
  'athlete_weight': '80.00000',
  'elevation_gain': '70.60000',
  'elevation_loss': '84.20000',
  'total_work': '527.35000',
  'average_speed': '33.19667',
  'average_po

In [11]:
i = 0
for sec in activities[n].metadata['METRICS']['power_mmp_secs']:
    if (sec == 1200):
        print(i)
        break
    i = i + 1

283


In [14]:
activities[n].metadata['METRICS']['power_mmp_secs'][:284]

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 185,
 190,
 195,
 200,
 205

## Plotting power curve of the activity

In [ ]:
df = pd.DataFrame({'duration':activities[n].metadata['METRICS']['power_mmp_secs'], 'power':activities[n].metadata['METRICS']['power_mmp']})
df.plot('duration', 'power', kind='scatter', logx=True)

In [ ]:
activities[n].data

## Plotting power duration of the activity

In [ ]:
activities[n].data.power.plot()

## Plotting heart rate duration of the activity

In [ ]:
activities[n].data.hr.plot()

## Plotting altitude duration of the training

### Btw, we can see that the athlete ride the same loop twice

In [ ]:
activities[n].data.alt.plot()

In [ ]:
activities[n].metadata['METRICS']['a_skiba_xpower'][0]

# Filter all activities with power data

In [ ]:
bike_activities = list()
for activity in activities:
    if 'power_mmp_secs' in activity.metadata['METRICS']:
        bike_activities.append(activity)
len(bike_activities)

# Plotting power curve for every activity for chosen athlete

In [ ]:
for activity in bike_activities:
    df = pd.DataFrame({'duration':activity.metadata['METRICS']['power_mmp_secs'], 'power':activity.metadata['METRICS']['power_mmp']})
    df.plot('duration', 'power', kind='scatter', logx=True)

# Create power curve based on all of activities for chosen athlete

In [ ]:
power_curve = {}
#acc = []
for activity in bike_activities:
    for i in range(len(activity.metadata['METRICS']['power_mmp_secs'])):
        if activity.metadata['METRICS']['power_mmp_secs'][i] in power_curve:
            power_curve[activity.metadata['METRICS']['power_mmp_secs'][i]] = max(power_curve[activity.metadata['METRICS']['power_mmp_secs'][i]],activity.metadata['METRICS']['power_mmp'][i])
        else:
            power_curve[activity.metadata['METRICS']['power_mmp_secs'][i]] = activity.metadata['METRICS']['power_mmp'][i]

In [ ]:
power_curve

In [ ]:
len(power_curve)

In [ ]:
powerDF = pd.DataFrame.from_dict(power_curve, orient='index')

In [ ]:
powerDF.reset_index()

In [ ]:
powerDF.plot(style='.',logx=True)

# Creating dataframe with interesting data of all activities for chosen athlete

In [59]:
local_athletes = list(od.local_athletes())

In [18]:
def month(df):
    return parse(df['date']).month

In [ ]:
def create_dataset(data):
    criticalPower = pd.DataFrame()
    titles = ['activity','athlete','date','weight','duration','coggan_tss','np']
    for activity in tqdm(data):
        rows = criticalPower
        s2 = pd.Series(activity.metadata['METRICS']['power_mmp'], index=activity.metadata['METRICS']['power_mmp_secs'])
        s1 = pd.Series([activity.id,athlete.id,activity.metadata['date'],activity.metadata['METRICS']['athlete_weight'],activity.metadata['METRICS']['time_riding'],activity.metadata['METRICS']['coggan_tss'],activity.metadata['METRICS']['coggan_np'][0]],index=titles)
        row = pd.concat([s1,s2])
        criticalPower = pd.concat([rows, row.to_frame().T], ignore_index=True)
    criticalPower['month'] = criticalPower.apply(lambda row: month(row), axis=1)
    return criticalPower


In [ ]:
PowerDF = create_dataset(bike_activities)
PowerDF

# Creating dataframe with power data of all activities for 120 athletes

In [19]:
def create_dataframe():
    criticalPower = pd.DataFrame()
    titles = ['activity','athlete','date','weight','duration','coggan_tss','np']
    errors = {}
    i = 0
    for athlete in tqdm(local_athletes):
        #athlete = local_athletes[idx]
        activities = list(athlete.activities())
        for activity in activities:
            try:
                if ('power_mmp_secs' in activity.metadata['METRICS'] and 'time_riding' in activity.metadata['METRICS'] and 'coggan_tss' in activity.metadata['METRICS']):
                    rows = criticalPower
                    s2 = pd.Series(activity.metadata['METRICS']['power_mmp'][:284], index=activity.metadata['METRICS']['power_mmp_secs'][:284])
                    s1 = pd.Series([activity.id,athlete.id,activity.metadata['date'],activity.metadata['METRICS']['athlete_weight'],activity.metadata['METRICS']['time_riding'],activity.metadata['METRICS']['coggan_tss'],activity.metadata['METRICS']['coggan_np'][0]],index=titles)
                    row = pd.concat([s1,s2])
                    criticalPower = pd.concat([rows, row.to_frame().T], ignore_index=True)
            except:
                errors[i] = [athlete, activity]
                i = i + 1
                continue
    try:
        criticalPower['month'] = criticalPower.apply(lambda row: month(row), axis=1)
        print(len(errors))
        print(errors)
        return criticalPower
    except:
        print(len(errors))
        print(errors)
        return criticalPower

In [20]:
PowerDF = create_dataframe()
PowerDF

100%|██████████| 120/120 [8:33:43<00:00, 256.86s/it]  


activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ...  960  990 1020 1050 1080 1110 1140 1170 1200 month  
0      451  435  427  ...  162  162  163  160  162  159  155  155  154     2  
1      778  760  725  ...  164  162  163  165  164  164  164  165  166     2  
2      535  532  521  ...  188  185  182  178  181  184  182  179  174     2  
3      553  529  509  ...  217  215  213  210  208  208  207  206  205     2  
4      382  381  374  ...  184  184  184  184  184  184  184  183  179     2  
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  
57586  278  261  253  ...  202  202  202  202  202  202  202  202  202     9  
57587  557  546  509  ...  224  224  227  229  231  229  227  225  223    10  
57588  248  235  228  ...  202  202  202  201  201  201  201  201  201    10  
57589  968  963  951  ...  201  201  200  199  199  198  197  196  194    10  
57590  929  929  927  ...  195  195  194  194  193  192  191  190  189    10  

[57591 rows x 292 columns]

# Function calculating ftp per month:

In [21]:
def calc_ftp(dataFrame):
    x = 0
    mon = 0
    athlete = ''
    monftp = []
    ftp = []
    acc = []
    for i, row in tqdm(dataFrame.iterrows()):
        if ((i + 1) == len(dataFrame)):
            if (np.isnan(row[1200])):
                acc = acc + [0]
            else:
                acc = acc + [row[1200]]
            x = x + 1
            monftp = [max(acc) * 0.95] * x
            ftp = ftp + monftp
        elif (i == 0):
            athlete = row['athlete']
            mon = row['month']
            if (np.isnan(row[1200])):
                acc = acc + [0]
            else:
                acc = acc + [row[1200]]
            x = x + 1
        elif (mon == row['month'] and athlete == row['athlete']):
            if (np.isnan(row[1200])):
                acc = acc + [0]
            else:
                acc = acc + [row[1200]]
            x = x + 1
        else:
            monftp = [max(acc) * 0.95] * x
            ftp = ftp + monftp
            x = 1
            monftp = []
        #print('i: ', i,'ftp: ', row[1200], ', month: ', row['month'], ', x: ', x)
        mon = row['month']
        athlete = row['athlete']
    return ftp

In [22]:
PowerDF['ftp'] = calc_ftp(PowerDF)
PowerDF

57591it [00:47, 1216.21it/s]


activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ...  990 1020 1050 1080 1110 1140 1170 1200 month  \
0      451  435  427  ...  162  163  160  162  159  155  155  154     2   
1      778  760  725  ...  162  163  165  164  164  164  165  166     2   
2      535  532  521  ...  185  182  178  181  184  182  179  174     2   
3      553  529  509  ...  215  213  210  208  208  207  206  205     2   
4      382  381  374  ...  184  184  184  184  184  184  183  179     2   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   
57586  278  261  253  ...  202  202  202  202  202  202  202  202     9   
57587  557  546  509  ...  224  227  229  231  229  227  225  223    10   
57588  248  235  228  ...  202  202  201  201  201  201  201  201    10   
57589  968  963  951  ...  201  200  199  199  198  197  196  194    10   
57590  929  929  927  ...  195  194  194  193  192  191  190  189    10   

          ftp  
0      242.25  
1      242.25  
2      242.25  
3      242.25  
4      242.25  
...       ...  
57586  428.45  
57587  428.45  
57588  428.45  
57589  428.45  
57590  428.45  

[57591 rows x 293 columns]

# Calculating power zones for every activity

Of course in one month power zones will not change

In [23]:
def calc_power_zones(dataFrame):
    athlete = ''
    mon = 0
    onePZ = []
    onePZ_acc = 0
    twoPZ = []
    twoPZ_acc = 0
    threePZ = []
    threePZ_acc = 0
    fourPZ = []
    fourPZ_acc = 0
    fivePZ = []
    fivePZ_acc = 0
    sixPZ = []
    sixPZ_acc = 0
    for i, row in tqdm(dataFrame.iterrows()):
        
        if (i == 0):
            athlete = row['athlete']
            mon = row['month']
            onePZ_acc = 0.55 * row['ftp']
            twoPZ_acc = 0.75 * row['ftp']
            threePZ_acc = 0.9 * row['ftp']
            fourPZ_acc = 1.05 * row['ftp']
            fivePZ_acc = 1.2 * row['ftp']
            sixPZ_acc = 1.5 * row['ftp']
            
            onePZ = onePZ + [onePZ_acc]
            twoPZ = twoPZ + [twoPZ_acc]
            threePZ = threePZ + [threePZ_acc]
            fourPZ = fourPZ + [fourPZ_acc]
            fivePZ = fivePZ + [fivePZ_acc]
            sixPZ = sixPZ + [sixPZ_acc]
            
        elif (mon == row['month'] and athlete == row['athlete']):
            onePZ = onePZ + [onePZ_acc]
            twoPZ = twoPZ + [twoPZ_acc]
            threePZ = threePZ + [threePZ_acc]
            fourPZ = fourPZ + [fourPZ_acc]
            fivePZ = fivePZ + [fivePZ_acc]
            sixPZ = sixPZ + [sixPZ_acc]
            
        else:
            onePZ_acc = 0.55 * row['ftp']
            twoPZ_acc = 0.75 * row['ftp']
            threePZ_acc = 0.9 * row['ftp']
            fourPZ_acc = 1.05 * row['ftp']
            fivePZ_acc = 1.2 * row['ftp']
            sixPZ_acc = 1.5 * row['ftp']
            
            onePZ = onePZ + [onePZ_acc]
            twoPZ = twoPZ + [twoPZ_acc]
            threePZ = threePZ + [threePZ_acc]
            fourPZ = fourPZ + [fourPZ_acc]
            fivePZ = fivePZ + [fivePZ_acc]
            sixPZ = sixPZ + [sixPZ_acc]
            
        mon = row['month']
        athlete = row['athlete']
    return onePZ, twoPZ, threePZ, fourPZ, fivePZ, sixPZ

In [24]:
PowerDF['1_power_zone'], PowerDF['2_power_zone'], PowerDF['3_power_zone'], PowerDF['4_power_zone'], PowerDF['5_power_zone'], PowerDF['6_power_zone'] = calc_power_zones(PowerDF)
PowerDF

57591it [01:10, 811.66it/s] 


activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ... 1170 1200 month     ftp 1_power_zone 2_power_zone  \
0      451  435  427  ...  155  154     2  242.25     133.2375     181.6875   
1      778  760  725  ...  165  166     2  242.25     133.2375     181.6875   
2      535  532  521  ...  179  174     2  242.25     133.2375     181.6875   
3      553  529  509  ...  206  205     2  242.25     133.2375     181.6875   
4      382  381  374  ...  183  179     2  242.25     133.2375     181.6875   
...    ...  ...  ...  ...  ...  ...   ...     ...          ...          ...   
57586  278  261  253  ...  202  202     9  428.45     235.6475     321.3375   
57587  557  546  509  ...  225  223    10  428.45     235.6475     321.3375   
57588  248  235  228  ...  201  201    10  428.45     235.6475     321.3375   
57589  968  963  951  ...  196  194    10  428.45     235.6475     321.3375   
57590  929  929  927  ...  190  189    10  428.45     235.6475     321.3375   

      3_power_zone 4_power_zone 5_power_zone 6_power_zone  
0          218.025     254.3625       290.70      363.375  
1          218.025     254.3625       290.70      363.375  
2          218.025     254.3625       290.70      363.375  
3          218.025     254.3625       290.70      363.375  
4          218.025     254.3625       290.70      363.375  
...            ...          ...          ...          ...  
57586      385.605     449.8725       514.14      642.675  
57587      385.605     449.8725       514.14      642.675  
57588      385.605     449.8725       514.14      642.675  
57589      385.605     449.8725       514.14      642.675  
57590      385.605     449.8725       514.14      642.675  

[57591 rows x 299 columns]

# Calculating time spent in every power zones for every activity

In [42]:
athletes_in_dataset = list(PowerDF['athlete'].unique())
athletes_in_dataset

['be3b662e-6fda-4c4f-85f8-812a364657f7',
 '705d4ac7-258c-42d6-9ff6-c74b3dbb04c7',
 '939eabb2-f1e4-4bce-8281-2baf58a8f772',
 '4727f484-f08d-48cd-b680-e3625aaa5d50',
 '810e205d-4e6c-4473-a5c4-a8b7568da315',
 '1bc1499d-9370-44a7-a2b8-cd1e8fabd8cd',
 '48a39786-5000-457c-b6d3-5f923e8f76a4',
 '6257f49d-0fa5-4eab-aa1f-7fe90c6aac67',
 'fa200051-3b4d-4509-b358-215094243f3e',
 'cf500e04-d5d4-428c-9106-8542720fe632',
 'adf7d4fc-8b3f-410c-81c2-7b713d659d1a',
 '9689a8d0-1cb1-42f6-83e4-fc0c8f6db31b',
 'c5245352-be7e-4382-b890-fa3283b1fd09',
 '9abb1be0-0f37-4014-ba36-a95568f2b105',
 'cf70e2a1-ad11-48dd-b836-265fadd0fcc2',
 '72bb0b9a-d866-4bca-a69b-89a93484d305',
 'ff64b289-b9d9-4927-8c7b-c4ba3a9945a1',
 'bd104cd6-e1c4-4724-967a-d3ee321d86a4',
 'ce043e2d-a873-49cb-a398-c6ce020877df',
 '6de3e82e-9d67-4c81-803e-88ccc69047a9',
 '4c08148d-096e-4593-8a67-2d08d4035df5',
 '42582e78-b367-42f3-9e79-8236c9abd526',
 '9e5e8829-e123-40cd-9f93-ab8385af4b0c',
 'f64ea63f-b74b-46a6-a677-8293ecf82daf',
 '2a95ee21-cde4-

In [47]:
path = '/Volumes/Mati_Data/data'
for folder_name in os.listdir(path):
    if folder_name not in athletes_in_dataset:
        full_folder_path = os.path.join(path, folder_name)
        st.rmtree(full_folder_path)

In [50]:
path = '/Volumes/Mati_Data/metadata'
for file_name in os.listdir(path):
    if file_name not in athletes_in_dataset:
        full_file_path = os.path.join(path, file_name)
        os.remove(full_file_path)

In [55]:
for athlete_id in tqdm(athletes_in_dataset):
    athlete = od.get_remote_athlete(athlete_id)
    athlete.store_locally(data=False)

100%|██████████| 119/119 [04:21<00:00,  2.19s/it]


In [56]:
for athlete_id in tqdm(athletes_in_dataset):
    local_athlete = LocalAthlete(athlete_id)
    local_athlete.download_remote_data()

100%|██████████| 119/119 [1:07:03<00:00, 33.81s/it]


In [67]:
ath = od.get_local_athlete('be3b662e-6fda-4c4f-85f8-812a364657f7')
activ = ath.get_activity('2018_02_14_12_54_23.csv')
activ.data['power']

0          0.0
1         53.0
2        100.0
3         92.0
4        117.0
         ...  
14036      0.0
14037      0.0
14038      9.0
14039      9.0
14040      0.0
Name: power, Length: 14041, dtype: float64

In [79]:
miss_files = {}
athlete_id = ''
athlete_path = ''
file_list = []
path = '/Volumes/Mati_Data/data'
iterator = 0
for i,row in tqdm(PowerDF.iterrows()):
    if (i == 0):
        athlete_id = row['athlete']
        athlete_path = os.path.join(path, athlete_id)
        file_list = [activ for activ in os.listdir(path)]
    elif (athlete_id != row['athlete']):
        athlete_id = row['athlete']
        athlete_path = os.path.join(path, athlete_id)
        file_list = [activ for activ in os.listdir(athlete_path)]
    if row['activity'] in file_list:
        pass
    else:
        iterator = iterator + 1
        miss_files[iterator] = [athlete_id, row['activity']]
print(len(miss_files))
miss_files

57591it [00:31, 1854.30it/s]

463


{1: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_14_12_54_23.csv'],
 2: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_15_10_55_37.csv'],
 3: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_16_11_27_08.csv'],
 4: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_17_09_11_17.csv'],
 5: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_18_06_29_10.csv'],
 6: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_18_08_23_07.csv'],
 7: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_19_06_35_49.csv'],
 8: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_20_09_43_33.csv'],
 9: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_23_16_13_07.csv'],
 10: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_24_16_14_11.csv'],
 11: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_25_11_17_20.csv'],
 12: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_25_15_26_16.csv'],
 13: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '2018_02_25_15_57_02.csv'],
 14: ['be3b662e-6fda-4c4f-85f8-812a364657f7', '

In [95]:
PowerDF = PowerDF.drop(463)

In [96]:
def time_in_zones(dataFrame):
    time_in_zone1 = []
    time_in_zone2 = []
    time_in_zone3 = []
    time_in_zone4 = []
    time_in_zone5 = []
    time_in_zone6 = []
    time_in_zone7 = []
    z1 = 0
    z2 = 0
    z3 = 0
    z4 = 0
    z5 = 0
    z6 = 0
    time_z1 = 0
    time_z2 = 0
    time_z3 = 0
    time_z4 = 0
    time_z5 = 0
    time_z6 = 0
    time_z7 = 0
    month = 0
    athlete_id = dataFrame['athlete'][0]
    athlete = od.get_local_athlete(athlete_id)
    #athlete_activities = list(athlete.activities())
    to_delete = {}
    for i, row in tqdm(dataFrame.iterrows()):
        if (i == 0):
            z1 = row['1_power_zone']
            z2 = row['2_power_zone']
            z3 = row['3_power_zone']
            z4 = row['4_power_zone']
            z5 = row['5_power_zone']
            z6 = row['6_power_zone']
            month = row['month']
        elif (z1 != row['1_power_zone'] or athlete_id != row['athlete'] or month != row['month']):
            z1 = row['1_power_zone']
            z2 = row['2_power_zone']
            z3 = row['3_power_zone']
            z4 = row['4_power_zone']
            z5 = row['5_power_zone']
            z6 = row['6_power_zone']
            month = row['month']
            if (athlete_id != row['athlete']):
                athlete_id = row['athlete']
                athlete = od.get_local_athlete(athlete_id)
                #athlete_activities = list(athlete.activities())
        activity = athlete.get_activity(row['activity'])
        #if activity in athlete_activities:
        try:
            for i, row in activity.data.iterrows():
                if (row['power'] <= z1 and row['power'] > 0):
                    time_z1 = time_z1 + 1
                elif (row['power'] <= z2 and row['power'] > z1):
                    time_z2 = time_z2 + 1
                elif (row['power'] <= z3 and row['power'] > z2):
                    time_z3 = time_z3 + 1
                elif (row['power'] <= z4 and row['power'] > z3):
                    time_z4 = time_z4 + 1
                elif (row['power'] <= z5 and row['power'] > z4):
                    time_z5 = time_z5 + 1
                elif (row['power'] <= z6 and row['power'] > z5):
                    time_z6 = time_z6 + 1
                elif (row['power'] > z6):
                    time_z7 = time_z7 + 1
            time_in_zone1 = time_in_zone1 + [time_z1]
            time_in_zone2 = time_in_zone2 + [time_z2]
            time_in_zone3 = time_in_zone3 + [time_z3]
            time_in_zone4 = time_in_zone4 + [time_z4]
            time_in_zone5 = time_in_zone5 + [time_z5]
            time_in_zone6 = time_in_zone6 + [time_z6]
            time_in_zone7 = time_in_zone7 + [time_z7]
            time_z1 = 0
            time_z2 = 0
            time_z3 = 0
            time_z4 = 0
            time_z5 = 0
            time_z6 = 0
            time_z7 = 0
        except:
            to_delete[i] = [row['athlete'], row['activity']]
            time_in_zone1 = time_in_zone1 + [0]
            time_in_zone2 = time_in_zone2 + [0]
            time_in_zone3 = time_in_zone3 + [0]
            time_in_zone4 = time_in_zone4 + [0]
            time_in_zone5 = time_in_zone5 + [0]
            time_in_zone6 = time_in_zone6 + [0]
            time_in_zone7 = time_in_zone7 + [0]
            time_z1 = 0
            time_z2 = 0
            time_z3 = 0
            time_z4 = 0
            time_z5 = 0
            time_z6 = 0
            time_z7 = 0
    print(len(to_delete))
    print(to_delete)
    return time_in_zone1, time_in_zone2, time_in_zone3, time_in_zone4, time_in_zone5, time_in_zone6, time_in_zone7

In [97]:
PowerDF['time_in_zone1'], PowerDF['time_in_zone2'], PowerDF['time_in_zone3'], PowerDF['time_in_zone4'], PowerDF['time_in_zone5'], PowerDF['time_in_zone6'], PowerDF['time_in_zone7'] = time_in_zones(PowerDF)
PowerDF


57590it [8:33:58,  1.87it/s]


0
{}


activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ... 5_power_zone 6_power_zone time_in_zone1  \
0      451  435  427  ...       290.70      363.375          5789   
1      778  760  725  ...       290.70      363.375          8903   
2      535  532  521  ...       290.70      363.375          3815   
3      553  529  509  ...       290.70      363.375          3676   
4      382  381  374  ...       290.70      363.375          1403   
...    ...  ...  ...  ...          ...          ...           ...   
57586  278  261  253  ...       514.14      642.675          5608   
57587  557  546  509  ...       514.14      642.675          2984   
57588  248  235  228  ...       514.14      642.675          6158   
57589  968  963  951  ...       514.14      642.675          1521   
57590  929  929  927  ...       514.14      642.675          1457   

      time_in_zone2 time_in_zone3 time_in_zone4 time_in_zone5 time_in_zone6  \
0              1614           776           437           199           351   
1              2609           924           630           231           363   
2               969           458           482           231           369   
3              1484          1254          1137           733           329   
4               351           486           258            98            33   
...             ...           ...           ...           ...           ...   
57586            33             0             0             0             0   
57587            10            14           281            18             4   
57588             9             0             0             0             0   
57589            47             1             1             3             6   
57590            78             1             0             3             5   

      time_in_zone7         tss  
0                79  139.436932  
1               127  156.936396  
2               132  110.518944  
3                70  135.683157  
4                 2   28.209429  
...             ...         ...  
57586             0   32.467762  
57587             0   25.777388  
57588             0   35.507297  
57589            23   11.243807  
57590            23   10.329913  

[57590 rows x 307 col

# TSS calculation function

In [98]:
def calc_tss(dataFrame):
    tss = []
    for i, row in tqdm(dataFrame.iterrows()):
        tss = tss + [(float(row['duration'])*float(row['np'])*float(row['np'])/float(row['ftp']))/(float(row['ftp'])*3600)*100]
    return tss

In [99]:
PowerDF['tss'] = calc_tss(PowerDF)
PowerDF

57590it [00:40, 1411.27it/s]


activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ... 5_power_zone 6_power_zone time_in_zone1  \
0      451  435  427  ...       290.70      363.375          5789   
1      778  760  725  ...       290.70      363.375          8903   
2      535  532  521  ...       290.70      363.375          3815   
3      553  529  509  ...       290.70      363.375          3676   
4      382  381  374  ...       290.70      363.375          1403   
...    ...  ...  ...  ...          ...          ...           ...   
57586  278  261  253  ...       514.14      642.675          5608   
57587  557  546  509  ...       514.14      642.675          2984   
57588  248  235  228  ...       514.14      642.675          6158   
57589  968  963  951  ...       514.14      642.675          1521   
57590  929  929  927  ...       514.14      642.675          1457   

      time_in_zone2 time_in_zone3 time_in_zone4 time_in_zone5 time_in_zone6  \
0              1614           776           437           199           351   
1              2609           924           630           231           363   
2               969           458           482           231           369   
3              1484          1254          1137           733           329   
4               351           486           258            98            33   
...             ...           ...           ...           ...           ...   
57586            33             0             0             0             0   
57587            10            14           281            18             4   
57588             9             0             0             0             0   
57589            47             1             1             3             6   
57590            78             1             0             3             5   

      time_in_zone7         tss  
0                79  139.436932  
1               127  156.936396  
2               132  110.518944  
3                70  135.683157  
4                 2   28.209429  
...             ...         ...  
57586             0   32.467762  
57587             0   25.777388  
57588             0   35.507297  
57589            23   11.243807  
57590            23   10.329913  

[57590 rows x 307 col

In [103]:
PowerDF

activity                               athlete  \
0      2018_02_14_12_54_23.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
1      2018_02_15_10_55_37.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
2      2018_02_16_11_27_08.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
3      2018_02_17_09_11_17.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
4      2018_02_18_06_29_10.csv  be3b662e-6fda-4c4f-85f8-812a364657f7   
...                        ...                                   ...   
57586  2018_09_30_10_02_16.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57587  2018_10_01_18_37_23.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57588  2018_10_08_10_10_11.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57589  2018_10_09_12_15_24.csv  44c54229-400f-4809-a0a6-5eb65fdad381   
57590  2018_10_09_12_15_43.csv  44c54229-400f-4809-a0a6-5eb65fdad381   

                          date    weight     duration coggan_tss         np  \
0      2018/02/14 03:54:23 UTC  55.00000  13961.00000  104.97136  145.25988   
1      2018/02/15 01:55:37 UTC  55.00000  18214.00000  174.00259  134.91934   
2      2018/02/16 02:27:08 UTC  55.00000   8921.00000  104.54728  161.78067   
3      2018/02/17 00:11:17 UTC  55.00000  10115.00000  140.12042  168.34300   
4      2018/02/17 21:29:10 UTC  55.00000   3286.00000   43.31742  134.67234   
...                        ...       ...          ...        ...        ...   
57586  2018/09/30 05:02:16 UTC  62.00000   5644.00000   76.44757  194.97731   
57587  2018/10/01 13:37:23 UTC  62.00000   3311.00000   60.04243  226.82518   
57588  2018/10/08 05:10:11 UTC  62.00000   6176.00000   83.78151  194.92007   
57589  2018/10/09 07:15:24 UTC  62.00000   1601.00000   25.98709  215.43310   
57590  2018/10/09 07:15:43 UTC  62.00000   1569.00000   24.37699  208.58749   

         1    2    3  ... 5_power_zone 6_power_zone time_in_zone1  \
0      451  435  427  ...       290.70      363.375          5789   
1      778  760  725  ...       290.70      363.375          8903   
2      535  532  521  ...       290.70      363.375          3815   
3      553  529  509  ...       290.70      363.375          3676   
4      382  381  374  ...       290.70      363.375          1403   
...    ...  ...  ...  ...          ...          ...           ...   
57586  278  261  253  ...       514.14      642.675          5608   
57587  557  546  509  ...       514.14      642.675          2984   
57588  248  235  228  ...       514.14      642.675          6158   
57589  968  963  951  ...       514.14      642.675          1521   
57590  929  929  927  ...       514.14      642.675          1457   

      time_in_zone2 time_in_zone3 time_in_zone4 time_in_zone5 time_in_zone6  \
0              1614           776           437           199           351   
1              2609           924           630           231           363   
2               969           458           482           231           369   
3              1484          1254          1137           733           329   
4               351           486           258            98            33   
...             ...           ...           ...           ...           ...   
57586            33             0             0             0             0   
57587            10            14           281            18             4   
57588             9             0             0             0             0   
57589            47             1             1             3             6   
57590            78             1             0             3             5   

      time_in_zone7         tss  
0                79  139.436932  
1               127  156.936396  
2               132  110.518944  
3                70  135.683157  
4                 2   28.209429  
...             ...         ...  
57586             0   32.467762  
57587             0   25.777388  
57588             0   35.507297  
57589            23   11.243807  
57590            23   10.329913  

[57590 rows x 307 col

# Machine Learning Model

task to do:
1. przygotować zbiór danych miesięczynych (jeden wiersz odpowiada jednemu miesiącowi treningu dla jedngo zawodnika)
2. dodać kolumnę z bezwzględną zmianą FTP oraz z procentową zmianą FTP
3. Przygotować dane do trenowania, testowania i wwalidacji
4. Stworzyć model drzewa do przewidywania zmiany FTP
5. Znaleźć sposób na znalezienie wartości maksymalnej zmiany FTP* oraz znaleźć metodę optymalizacji parametrów wejściowych (głównie czas w strefach i TSS) tak aby uzyskać maksymalny wzrost FTP.

* tu pojawia się wątpliwość, bo dla zawodników mniej wytrenowanych wzrost FTP będzie łatwiejszy i bardziej wyraźny, jednocześnie nie oznacza to, że trenują lepiej od zawodników zaawansowanych. Dlatego pytanie, czy zawodnicy z niższym FTP nie "zagłuszą" zawodników z wyższym.